oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java

```diff
@@ -142,6 +233,15 @@ private static OIDCProviderMetadata parse(String body) {
 		}
 	}
 
+	private static AuthorizationServerMetadata parseAuthorizationServerMetadata(String body) {
```

Comment:
Yeah, these checked exceptions make the code a little noisy. You might consider:

```java
private interface ThrowingFunction<S, T, E extends Throwable> {
    T apply(S src) throws E;
}
```

And then:

```java
private static <T> T parse(String body, ThrowingFunction<String, T, ParseException> parser) {
    try {
        return parser.apply(body);
    } catch (ParseException e) {
        throw new RuntimeException(e);
    }
}
```

It's probably sixes, though it could help clean up the two public methods:

```java
public static ClientRegistration.Builder withOidcIssuerLocation(String issuer) {
    String configuration = getOpenIdConfiguration(issuer);
    OIDCProviderMetadata metadata = parse(configuration, OIDCProviderMetadata::parse);
    return withProviderConfiguration(metadata)
            .userInfoUri(metadata.getUserInfoEndpointURI().toASCIIString());
}

public static ClientRegistration.Builder withOAuth2IssuerLocation(String issuer) {
    String configuration = getOAuth2Configuration(issuer);
    AuthorizationServerMetadata metadata = parse(configuration, AuthorizationServerMetadata::parse);
    return withProviderConfiguration(metadata);
}

private static ClientRegistration.Builder withProviderConfiguration(AuthorizationServerMetadata metadata) {
    // ... everything except for user info uri
}
```

Reply:
That definitely simplifies the public methods. 
We might have to pass the `issuer` as well to `withProviderConfiguration`. I also had to update the error message for `issuer` mismatch to as follow 

```
throw new IllegalStateException("The Issuer \"" + metadataIssuer + "\" provided in the configuration metadata did "
					+ "not match the requested issuer \"" + issuer + "\"");
```

Reply:
That's correct. `withOAuth2IssuerLocation` was just a typo on my part.